src/librustc_typeck/check/mod.rs

```diff
@@ -4574,10 +4582,23 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
         // Add all the obligations that are required, substituting and
         // normalized appropriately.
-        let bounds = self.instantiate_bounds(span, def_id, &substs);
-        self.add_obligations_for_parameters(
-            traits::ObligationCause::new(span, self.body_id, traits::ItemObligation(def_id)),
-            &bounds);
+        let (bounds, spans) = self.instantiate_bounds(span, def_id, &substs);
+
+        for (i, mut obligation) in traits::predicates_for_generics(
```

Comment:
This method is nearly 300 LOC now -- can you refactor this into its own method to not make things worse?
(and the method should really be split up more!)

Reply:
Pulling this code out. I'll leave splitting `instantiate_value_path` further to a follow up PR.